### PR TITLE
[cfengine] Initial integration

### PR DESCRIPTION
**Pull request by AdamKorcz**
This PR adds initial integration of [cfengine](https://github.com/cfengine/core).

Users include ([source](https://cfengine.com/)):
- Deutsche Telekom
- Panasonic
- Samsung
- Linkedin
- Pfizer
- DHL

----------------------------------------------------

@olehermanse @vpodzime 
I have been working on integrating cfengine into OSS-fuzz to take the first step with continuous fuzzing and have set up this PR with the build files and a simple fuzzer implemented by way of LibFuzzer.

OSS-fuzz is a free service offered by Google to run the fuzzers of critical open source projects free of charge. If bugs are found, maintainers on the mailing list get notified with a link to a detailed bug report. The service is offered with an implied expectation that bugs are found.

To finish the integration at least one maintainers email address is needed on the OSS-fuzz side.

Let me know if you are interested in proceeding with integration.

### CODE REVIEW

**Comment by vpodzime** [comment]
@AdamKorcz, I think this is a great initiative, thanks! Feel free to use my address: vratislav.podzimek@northern.tech.

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [inline projects/cfengine/string_fuzzer.c:25]: How/why did you choose `ParseStringExpression()`?

We've done some fuzzing before, manually, but then we've fuzzed the complete parser:

https://github.com/olehermanse/dotfiles/blob/master/cfengine/fuzz.sh

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [inline projects/cfengine/string_fuzzer.c:24]: Can `data` have zero-bytes? Is it possible to instruct the fuzzer to not insert zeroes, as that might be an assumption at this point?

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [inline projects/cfengine/string_fuzzer.c:18]: How would we add fuzzing of more functions in this setup?

File: projects/cfengine/build.sh
Comment (olehermanse) [inline projects/cfengine/build.sh:26]: Is there something special about the name `string_fuzzer` or will any programs in `$OUT/` be ran?

File: projects/cfengine/project.yaml
Comment (olehermanse) [inline projects/cfengine/project.yaml:10]: What happens if the fuzzed code fails an `assert()`?

**Review by olehermanse — COMMENTED** [review]
Thanks for working on this! I just have some questions about how it works.

File: projects/cfengine/string_fuzzer.c
Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c:25]: Hitting the complete parser is a good goal, and doable as well. I will be glad to give it a go, but I would prefer this to be a separate endeavour once CFengine is up and running as a it requires a bit more effort than I have available at this moment.

File: projects/cfengine/string_fuzzer.c
Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c:18]: A single file can maximum have one LLVMFuzzerTestOneInput api in it. There are several options to add more fuzzers, of which one is to have several separate files, or if the input is similar we could use a single LLVMFuzzerTestOneInput to call into several targets.

File: projects/cfengine/project.yaml
Comment (AdamKorcz) [reply projects/cfengine/project.yaml:10]: `exit`s and `assert`s are fuzz-blockers and need to be rewritten. I suggest starting with the least-intrusive way of doing this, and then gradually improve the fuzzability. A good starting point could be to define a "fuzz"-macro in the makefile that avoids `exit`s and `assert`s.

File: projects/cfengine/build.sh
Comment (AdamKorcz) [reply projects/cfengine/build.sh:26]: The compiles binaries need to be placed in `$OUT/`. They can be called anything.

File: projects/cfengine/string_fuzzer.c
Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c:24]: Yes, LibFuzzer (the fuzzing engine used for this harness) can generate zero-bytes.

It is also possible to instruct the harness to not add those to the input.

Why would we need to do that?

Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c:24]: We are testing an internal function / an internal part of the parser. It might be the case that this string has already been checked for zero-bytes, and we don't want this specific code to re-check / handle that case. I'm not too familiar with this part of the codebase, but I would guess that this is the case here, that the parser will never send a string with a zero to this function. (But there is no assert there, so I don't really know this).

I want the fuzzer to find issues which can happen in the real world, i.e. either test the full parser, or test internal parts, but with some restrictions depending on the context.

So, I would suggest to start with instructing libfuzz to not generate zero-bytes in this case. In the end, I think it will lead to finding more important problems.

If we test the full parser, we can definitely have the fuzzer insert zeroes as well.

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c:25]: Sure, no problem, I was more interested in an answer as to why you chose this part.

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c:18]: Thanks for the answer, make ssense.

File: projects/cfengine/project.yaml
Comment (olehermanse) [reply projects/cfengine/project.yaml:10]: Can you elaborate on what you mean by fuzz-blockers?

We don't want to remove / rewrite asserts, they are quite valuable.

On the other hand, the fuzzer should never fail an assert, if it does that's a bug (given that we are generating appropriate inputs for the context). That's primarily why I ask. If it hits an assert which fails, do we get a bug report?

File: projects/cfengine/build.sh
Comment (olehermanse) [reply projects/cfengine/build.sh:26]: Oki, thanks.

File: projects/cfengine/string_fuzzer.c
Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c:24]: Thank you for elaborating. I will just keep this draft open and check during the week if we can hit the full parser in this iteration before merging. 
Before I do that, could you confirm that you are interested in fuzzing CFEngine continuously?

Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c:24]: @AdamKorcz Yes we are interested.

Starting with a small target like this makes a lot of sense, I don't think you have to or should set up fuzzing of the full parser, but I think for an internal function like this, we should add the limitation of no zero-bytes, as explained above.

Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c:24]: That sounds good, thank you for the feedback. I will amend the fuzzer to prevent zero-bytes being passed to the target.

File: projects/cfengine/project.yaml
Comment (olehermanse) [reply projects/cfengine/project.yaml:10]: @AdamKorcz I don't think this comment is resolved, I'm still interested to know if we receive bug reports for failed asserts (I hope so).

File: projects/cfengine/string_fuzzer.c
Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c]: @olehermanse I updated the fuzzer. Was that what you had in mind?

Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c]: I don't know what the return code (`1`) means here, but I guess that will work. Have you tested it?

Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c]: It worked fine locally but I changed the fuzzer to return `0` in all conditions.

Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c]: @olehermanse Kind reminder.

Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c]: @AdamKorcz You didn't answer the other comment, I asked whether we will get bug reports if it hits an assert (and it fails).

Comment (AdamKorcz) [reply projects/cfengine/string_fuzzer.c]: Apologies. 

> I asked whether we will get bug reports if it hits an assert (and it fails)

In this case you will get bug reports.

Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c]: Thanks for the clarification.

**Review by olehermanse — APPROVED** [review]
LGTM, thanks for your work on this 🚀

**Comment by AdamKorcz** [comment]
@inferno-chromium @jonathanmetzman @oliverchang @asraa @DavidKorczynski Ready for review.

**Comment by jonathanmetzman** [comment]
Started our internal review process for this.

File: projects/cfengine/string_fuzzer.c
Comment (jonathanmetzman) [inline projects/cfengine/string_fuzzer.c:23]: I wonder if it would be better to change all 0s to some other constant.
A downside would be that testcases would still have 0s making them harder for maintainers to debug (they would have to know that they get changed to some other constant or use this target.

Hopefully it shouldn't matter but it would be interesting (from a fuzzer author point of view) to see how many tests get thrown out by this check. It seems like it would be hard to generate a string that say needed to be 500 chars without having any 0s (1-1/256)**500=0.14

**Review by jonathanmetzman — APPROVED** [review]
LGTM

File: projects/cfengine/string_fuzzer.c
Comment (olehermanse) [reply projects/cfengine/string_fuzzer.c:23]: Yea, I guess replacing with a constant would be better, or even better, getting the random number generator to go from 1-255.

**Pull request merged**
